### DISCUSSION TITLE
multiple servers = problem

### DISCUSSION THREAD

**Discussion by yuriy23** [eModbus/eModbus#313] [Q&A] [answered]
there is a problem when two servers are working at once. here are the code and errors. how to do it correctly when there are several servers ? Thank you.

```.[1;33m[E] 455| ModbusServerRTU.cpp  [ 252] serve: RTU receive: E2 - CRC check error   ```
(but it is not always there, one for 20-100 requests.)

```#include <Arduino.h>
//#include <main.h>
//#include "plc_conf.h"
#include "HardwareSerial.h"
#include "ModbusClientRTU.h"
#include "ModbusServerRTU.h"
#include "CoilData.h"

#include <WiFi.h>

// Modbus server TCP
#include "ModbusServerTCPasync.h"

#ifndef MY_SSID
#define MY_SSID "_2.4GHz"
#endif
#ifndef MY_PASS
#define MY_PASS "76786723"
#endif

//Modbus 1
#define RXPIN_1 GPIO_NUM_18 //GPIO_NUM_18
#define TXPIN_1 GPIO_NUM_16 //GPIO_NUM_16
#define REDEPIN_1 GPIO_NUM_17 //GPIO_NUM_17
#define BAUDRATE 57600
//Modbus 2
#define RXPIN_2 GPIO_NUM_10 //GPIO_NUM_18
#define TXPIN_2 GPIO_NUM_8 //GPIO_NUM_16
#define REDEPIN_2 GPIO_NUM_9 //GPIO_NUM_17
#define BAUDRATE 57600

uint16_t mi_16BitBuf[16000];

#define BAUD_RATE 115200 // основной порт программы RW

char ssid[] = MY_SSID;                     // SSID and ...
char pass[] = MY_PASS;                     // password for the WiFi network used

//ModbusClientRTU MBslave1(REDEPIN_1);
ModbusServerRTU MBserver1(2000,REDEPIN_1);
//ModbusClientRTU MBslave2(REDEPIN_2);
ModbusServerRTU MBserver2(2000,REDEPIN_2);

// Create server
ModbusServerTCPasync MBserverTCP;
ModbusMessage FC03(ModbusMessage request);

void setup() {
Serial.begin(BAUD_RATE);
Serial1.begin(57600, SERIAL_8N1, RXPIN_1, TXPIN_1);
Serial2.begin(57600, SERIAL_8N1, RXPIN_2, TXPIN_2);

// Connect to WiFi
  WiFi.begin(ssid, pass);
  delay(200);
  while (WiFi.status() != WL_CONNECTED) {
    Serial.print(". ");
    delay(500);
  }
  IPAddress wIP = WiFi.localIP();
  Serial.printf("WIFi IP address: %u.%u.%u.%u\n", wIP[0], wIP[1], wIP[2], wIP[3]);

MBserver1.registerWorker(0x01, READ_HOLD_REGISTER, FC03);
MBserver2.registerWorker(0x02, READ_HOLD_REGISTER, FC03);
MBserverTCP.registerWorker(0x05, READ_HOLD_REGISTER, FC03);
MBserver1.begin(Serial1);
MBserver2.begin(Serial2);
//MBserverTCP.start(502, 1, 20000);
}
void loop() {

}

  //************************************************************************************************
// FC03: задача  с функциональным кодом 0x03 (READ_HOLD_REGISTER)
ModbusMessage FC03(ModbusMessage request) {
  uint16_t address;        // запрашиваемый  адрес
  uint16_t num;            // запрашиваемое количество регистров
  
  ModbusMessage response;  // ответное сообщение, которое будет отправлено обратно
  // получить значения запроса
  request.get(2, address);
  request.get(4, num);
	 if ( address<10000)     
	 {	
    //Настроить сообщение с идентификатором сервера, FC и длиной данных
    response.add(request.getServerID(), request.getFunctionCode(), (uint8_t)(num * 2));
    // Заполнить ответ запрошенными данными...
    for (uint16_t i = address; i < address + num; ++i) {
      response.add(mi_16BitBuf[i]);
      Serial.print("mi_16BitBuf[i]= ");Serial.println(mi_16BitBuf[i]);
    }
  } else{
   // Нет, ни адрес, ни количество не допускаются. Настройка ответа на ошибку.
    response.setError(request.getServerID(), request.getFunctionCode(), ILLEGAL_DATA_ADDRESS);
  }
  return response;
}
//************************************************************************************************```

one at a time , it works fine...

**Comment by Miq1** [comment]
Should work... I am doing that myself without any problem. 

[edited - saw the error line only now] The notorious intermittent CRC errors mostly are caused by wiring issues, or the older UART handling.  You may need to call ``prepareHardwareSerial()`` in ``setup()`` to properly init it, as is described [here in the docs](https://emodbus.github.io/modbusserver-rtu-api#important-when-using-hardwareserial)

Please note that both server tasks will call ``FC03``, which may lead to a race condition when both are accessing ``mi_16BitBuf``.

**Comment by yuriy23** [comment]
Thanks Mik1. I read it, but I can't execute it, I get compiler errors. although arduino > 2.00  what am I doing wrong ?

![2023-09-06_09-50-47](https://github.com/eModbus/eModbus/assets/5504086/81e58135-2221-4c00-b37b-da43c6ac725b)
![2023-09-06_09-49-50](https://github.com/eModbus/eModbus/assets/5504086/94cb00e5-2ff8-400b-9016-71ed9de83d8f)
![2023-09-06_09-49-25](https://github.com/eModbus/eModbus/assets/5504086/1c76de98-4c3c-497b-a512-9db4bd474d35)

**Comment by Miq1** [comment]
Did you download the latest eModbus lib?

yuriy23 ***@***.***> schrieb am Mi., 6. Sept. 2023, 08:56:

**Reply by yuriy23** [reply]
Yes
![2023-09-06_11-19-51](https://github.com/eModbus/eModbus/assets/5504086/bbe31662-6753-4961-bb11-ea9f4c51b5ee)

maybe the problem is in the arduino 2.0.11 framework ?

**Comment by Miq1** [comment, marked as answer]
Since it is a static method of RTUutils, you need to use ``::`` instead of
``.``.

yuriy23 ***@***.***> schrieb am Mi., 6. Sept. 2023, 10:21:

**Reply by yuriy23** [reply]
but it didn 't help with CRC . it doesn't work together anyway, individually ok.

**Reply by yuriy23** [reply]
https://youtu.be/xLttx2qrs04

**Comment by Miq1** [comment]
It would be interesting to see the raw data. Can you compile with ``-DLOG_LEVEL=6`` and show the output here?

**Comment by yuriy23** [comment]
[Новый текстовый документ (3).txt](https://github.com/eModbus/eModbus/files/12536480/3.txt)

**Reply by yuriy23** [reply]
I understand the problem here. the data is not separated... Or am I wrong?


[V] 2480| RTUutils.cpp         [ 271] receive: S/[V] Raw buffer received: @3FCB553C/33:
  | 0000: 02 03 14 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
  | 0010: 00 00 00 00 00 00 00 F7  82 02 03 00 01 00 0A 94  |................|
  | 0020: 3E                                                |>               |
[D] 2511| RTUutils.cpp         [ 436] receive: S/[D] Received packet: @3FCAE664/1:
  | 0000: E2                                                |.               |
.[1;33m[E] 2521| ModbusServerRTU.cpp  [ 252] serve: RTU receive: E2 - CRC check error

**Reply by Miq1** [reply]
Right. That is caused by the client sending requests faster than the Modbus RTU quiet time is requiring - 3.5 character lengths minimum (that is about the time it takes to transmit 35 bits, hence 35000000/57600=608us theoretically, but the standard is requiring 1700us as a minimum). Most of the time it is working with a few microseconds above that limit, but sometimes it fails it.

**Reply by yuriy23** [reply]
ok. I will measure the hardware, then I will post the result

**Reply by yuriy23** [reply]
the problem is soft. this happens only if 2 uart channels are enabled one at a time, everything is fine  even if the scan time is 2ms. when two channels are turned on, the problem begins if the scan time is less than 150ms. there is a pause of about 140-160ms between the request and the response, which is a lot. I can see it on the oscilloscope.


![2023-09-06_14-56-28](https://github.com/eModbus/eModbus/assets/5504086/d8e27d60-72f7-4075-a46f-918169caab32)

**Reply by yuriy23** [reply]
sorry for my english

**Reply by Miq1** [reply]
The only thing I can imagine is a FreeRTOS one: you are employing 4 to 5 tasks (``loop``, two RTU server worker tasks, a TCP server task and maybe TCP client tasks). I have no idea what the task switches will take in terms of time.
Both RTU servers have their log output intermixed unfortunately, so a look at the individual timings is slightly blurred. One day I may add the task number to the log output...

**Reply by yuriy23** [reply]
I left one server and nothing else. then he added a second one. I hope this helps.

https://youtu.be/MrMfV20-piM?si=U8nL1T8zji2Zj03O

**Reply by Miq1** [reply]
Looks to me like the time between requests seems not to affect the error rate very much. Can you count the errors and the requests and calculate the ratio over some time? If it stays roughly the same for different scan rates, we may have a logic issue. If it is related to the scan rate, it rather will be a resource problem.

**Reply by yuriy23** [reply]
it does not separate two UARTs, but puts them in a queue and simultaneously connects data from different uarts into one buffer. IMHO
accordingly , 3.5 characters does not work . it is necessary to make it clear where the data is falling from which uart. 
I'm not a programmer, I could be wrong. I'm an electronics engineer.
[https://www.youtube.com/watch?v=glpbWtvlhRc](url)

**Reply by yuriy23** [reply]
good night

**Comment by Miq1** [comment]
> it does not separate two UARTs, but puts them in a queue and simultaneously connects data from different uarts into one buffer.

That would be disastrous! I cannot believe this really is happening - else the arduino-esp32 core developers would be drowning in issue reports.

In the lib, each RTU server is completely independent from others - different tasks, different local memory etc. So a mixture of two UARTs is not very likely as well.

There still are two potential race conditions in your code. Both servers may try to access your data area at the same time and both may concurrently try to print log lines to ``Serial``. The latter is no real issue but just may mix up the log lines.

Aside from that, I am still wondering why you have two RTU buses. You also are using different server IDs on the buses - why at all? And how do you send the requests to both of the servers? In your videos only one PC client is to be seen, not two, as would be expectable.

**Comment by yuriy23** [comment]
good morning. the device that I'm doing has 2 modbus and maybe there will be  TCP. 2 servers may be needed if you connect an HMI to one, and a plc to the second, which also polls this device. I have connected it in parallel for now because the second USB-485 is in another place, I will go for it on Saturday.

**Reply by yuriy23** [reply]
![2023-09-07_09-24-22](https://github.com/eModbus/eModbus/assets/5504086/027e54b7-a9a8-4eca-8e72-edd4cb7f6726)
if I'm not mistaken, this says that it all lies in one buffer

**Reply by Miq1** [reply]
What does "I have connected it in parallel for now" mean? Did you attach both RS485 adapters to one RTU bus? Interesting, never tried that myself. One effect from that will be to have both servers process the same data at the same time on the same core. This will require some heavy task switching, as the UART interrupts will fire for both tasks simultaneously.

And right, previous response of server ID 2 and the next request to it were ending up in the same buffer, namely the receive buffer of server ID 1, that is to ignore both, but did not get the processing time to notice the gap between - hence a single buffer.

You could try to use the optional core ID parameter when initializing the servers to have both run on different cores. This will affect other tasks, but may improve the CRC errors.

**Comment by yuriy23** [comment]
I took the second USB-RS485 now each server is connected to its master. there are 20 errors for about 20,000 requests. apparently there are still collisions, but there are not so many of them

**Reply by Miq1** [reply]
Glad to read that. Since the PC does multi-tasking now as well, the requests will not be simultaneously every time (as seen from the ESP32), plus, the respective RS485 bus will only see the requests for the attached server - no need any more to skip requests and responses for other servers as well. So the problem should be reduced to task priority clashes now when both server tasks are to process a request.
I would consider the error rate of 0.1% tolerable now.

**Reply by yuriy23** [reply]
I agree. I added another TCP server, just for the sake of interest, there were several more errors. but also not critical. Thank you, I think the topic can be closed. have a good weekend.

**Comment by yuriy23** [comment]
![0-02-05-ea60072da628f4d765bc5d23c8be238eb2cee74bf71b080fead8d730ebf2abc6_1e34e36e39224f81](https://github.com/eModbus/eModbus/assets/5504086/d120f5be-742c-4bfa-a896-d22e698e1c9e)

**Reply by yuriy23** [reply]
![0-02-05-324836bf1f2afbacaa3689ac621d7cd8e36fdf98982a3a421ba6438e9510227f_de9ff89add9c46c8](https://github.com/eModbus/eModbus/assets/5504086/53220bda-abfe-4f0a-87b8-04c4818b3e76)

**Reply by yuriy23** [reply]
two separate modbus channels

**Comment by Miq1** [comment]
👍🏻
